src/pwm.rs

```diff
@@ -0,0 +1,200 @@
+                type Duty = u16;
+
+                fn disable(&mut self, channel: Self::Channel) {
+                    match channel.0 {
+                        0 if self.ch0.is_some() => self.timer.chctl2.modify(|_r, w| w.ch0en().clear_bit()),
+                        1 if self.ch1.is_some() => self.timer.chctl2.modify(|_r, w| w.ch1en().clear_bit()),
+                        2 if self.ch2.is_some() => self.timer.chctl2.modify(|_r, w| w.ch2en().clear_bit()),
+                        3 if self.ch3.is_some() => self.timer.chctl2.modify(|_r, w| w.ch3en().clear_bit()),
+                        _ => (),
+                    }
+                }
+
+                fn enable(&mut self, channel: Self::Channel) {
+                    match channel.0 {
+                        0 if self.ch0.is_some() => self.timer.chctl2.modify(|_r, w| w.ch0en().set_bit()),
+                        1 if self.ch1.is_some() => self.timer.chctl2.modify(|_r, w| w.ch1en().set_bit()),
+                        2 if self.ch2.is_some() => self.timer.chctl2.modify(|_r, w| w.ch2en().set_bit()),
+                        3 if self.ch3.is_some() => self.timer.chctl2.modify(|_r, w| w.ch3en().set_bit()),
+                        _ => (),
+                    }
+                }
+
+                fn get_period(&self) -> Self::Time {
+                    return self.period;
+                }
+
+                fn get_duty(&self, channel: Self::Channel) -> Self::Duty {
+                    self.duty[channel.0 as usize]
+                }
+
+                fn get_max_duty(&self) -> Self::Duty {
+                    self.max_duty_cycle
+                }
+
+                fn set_duty(&mut self, channel: Self::Channel, duty: Self::Duty) {
+                    let mut duty = duty;
+                    if duty > self.max_duty_cycle {
+                        duty = self.max_duty_cycle
+                    }
+                    self.duty[channel.0 as usize] = duty;
+                    self.disable(channel.clone());
+                    match channel.0 {
+                        0 if self.ch0.is_some() => self.timer.ch0cv.write(|w| unsafe { w.bits(duty as $cvtype) }),
+                        1 if self.ch1.is_some() => self.timer.ch1cv.write(|w| unsafe { w.bits(duty as $cvtype) }),
+                        2 if self.ch2.is_some() => self.timer.ch2cv.write(|w| unsafe { w.bits(duty as $cvtype) }),
+                        3 if self.ch3.is_some() => self.timer.ch3cv.write(|w| unsafe { w.bits(duty as $cvtype) }),
+                        _ => (),
+                    }
+                    self.enable(channel);
+                }
+
+                fn set_period<P>(&mut self, period: P) where
+                    P: Into<Self::Time> {
+                    self.timer.ctl0.modify(|_, w| w.cen().clear_bit());
+                    self.timer.cnt.reset();
+
+                    let freq = period.into();
+
+                    let timer_clock = self.clocks.sysclk().0;
```

Comment:
Hmm. Seems like timers use wrong frequency source. I need to check this on my hardware.

Reply:
You're right, I missed that part from the datasheet and didn't realize TIMER0 has a different clock too. I'll send a PR, it's a 3 line change to select the correct clock.